### PR TITLE
Add latency stats for riak pb client operations [JIRA: RCS-243]

### PR DESCRIPTION
**Pull request by shino**
Part 3 of stats improvement which addresses #961 (RCS-37) (RCS-217) (RCS-217) (RCS-37) (RCS-37) (RCS-217) partially.

This PR adds latency/counter stats for (almost) every riak pb client operations.

Excluded operations and reasons are as follows:
- Ignored because not in hot path

```
  ./riak_cs_app.erl:115:    case catch riakc_pb_socket:get_bucket(MasterPbc, Bucket) of
```
- Ignored because strongly deprecated

```
./riak_cs_list_objects_fsm.erl:308:    riakc_pb_socket:stream_list_keys(ManifestPbc,
./riak_cs_list_objects_fsm.erl:496:    riakc_pb_socket:mapred_stream(ManifestPbc,
```

---

Also this PR fixes a bug that timeout is not properly set from
application environment for multipart upload listing call. Please
refer the commit message of dc637e1c9db1e5ee0c7086025a6f479a02183895
for details.

### CODE REVIEW

File: src/riak_cs_block_server.erl [hunk 1]
```diff
@@ -311,18 +308,19 @@ handle_local_notfound(RcPid, FullBucket, FullKey, GetOptions2,
...
-    case get_block_local(RcPid, FullBucket, FullKey, GetOptions2, Timeout) of
+    case get_block_local(RcPid, FullBucket, FullKey, GetOptions2, Timeout,
+                         [riakc, get_block_n_all]) of
         {ok, _} = Success ->
             ProceedFun(Success);
 
         {error, Why} when Why == disconnected;
                           Why == timeout ->
-            _ = lager:debug("get_block_local/5 failed: {error, ~p}", [Why]),
+            _ = lager:debug("get_block_local() failed: {error, ~p}", [Why]),
```
Comment (kuenishi) [inline src/riak_cs_block_server.erl:318]: Why is this changed to `()` ? Just `/6` looks nice.

Comment (shino) [inline src/riak_cs_block_server.erl:318]: The arity does not have information. I like log to be minimal but sufficient.

Comment (shino) [inline src/riak_cs_block_server.erl:318]: To be honest, I want to use standard or de-fact notation to refer a function that does not specify (or emphasize) its arity. Sometimes, `function_name/*` is used in Erlang, but it seems like a group of functions with the name (right?).
Any recommendation is appreciated :sweat_smile: 

Parenthesis with empty args may be not familiar.. Hmm... I saw the notation in exometer_core README.

File: src/riak_cs_gc.erl [hunk 2]
```diff
@@ -380,7 +380,8 @@ mark_manifests(RiakObject, Bucket, Key, UUIDsToMark, ManiFunction, RcPid) ->
     %% again without having to re-retrieve the object
     {ok, ManifestPbc} = riak_cs_riak_client:manifest_pbc(RcPid),
     riak_cs_pbc:put(ManifestPbc, UpdObj, [return_body],
-                    riak_cs_config:put_gckey_timeout()). %% <= bug: put_manifest_timeout should be used here
+                    riak_cs_config:put_gckey_timeout(), %% <= bug: put_manifest_timeout should be used here
+                    [riakc, put_manifest]).
```
Comment (kuenishi) [inline src/riak_cs_gc.erl:384]: These lines will conflict with #1184 . Will rebase #1184.

File: src/riak_cs_pbc.erl [hunk 3]
```diff
@@ -22,24 +22,65 @@
...
+        pong ->
+            _ = riak_cs_stats:update_with_start(StatsKey, StartTime);
+        _ ->
+            _ = riak_cs_stats:update_error_with_start(StatsKey, StartTime)
+    end,
+    Result.
+
 %% @doc Get an object from Riak
--spec get_object(pid(), binary(), binary(), proplists:proplist()|timeout()) ->
+-spec get_no_stats(pid(), binary(), binary(), proplists:proplist(), timeout()) ->
```
Comment (kuenishi) [inline src/riak_cs_pbc.erl]: I would prefer `get_with_no_stats` or `get_wo_stats` to this, just a grammar.

**Comment by kuenishi** [comment]
What is `include/riak_cs_stats.hrl` for? To include from multibag?

File: src/riak_cs_pbc.erl (hunk 3 above)
Comment (shino) [inline src/riak_cs_pbc.erl]: After the discussion in chat, `s/_no_stats/_sans_stats/`.

 Note for self: don't forget replace repair script under priv/.

File: src/riak_cs_riak_client.erl [hunk 4]
```diff
@@ -113,6 +114,36 @@ pbc_pool_name(undefined) ->
 pbc_pool_name(BagId) when is_binary(BagId) ->
     list_to_atom(lists:flatten(io_lib:format("pbc_pool_~s", [BagId]))).
 
+%% @doc Make a thunk that looks up samples for a given bucket+prefix.
```
Comment (kuenishi) [inline src/riak_cs_riak_client.erl]: s/prefix/postfix/ or s/prefix/suffix/ ?

File: src/riak_cs_riak_client.erl [hunk 5]
```diff
@@ -113,6 +114,36 @@ pbc_pool_name(undefined) ->
...
+-spec rts_puller(riak_client(), binary(), iolist(), riak_cs_stats:key()) -> fun().
+rts_puller(RcPid, Bucket, Postfix, StatsKey) ->
```
Comment (kuenishi) [inline src/riak_cs_riak_client.erl]: Looks like `Suffix` is major (Japanese page => http://blog.livedoor.jp/lalha/archives/50165797.html )

Comment (shino) [inline src/riak_cs_riak_client.erl]: I do have neither strong opinion nor enough knowledge. ... so consulted google-sensei and found the page [1]
In short, they are almost inter-changable but second answer said suffix is used as noun and postfix is as adjective. This makes sense for me. Then, the issue is what `Postfix` here means, whether short for "postfix token" or "suffix". After at length to here, I don't care which to be used and respect your choice :smile: 

[1] http://english.stackexchange.com/questions/81263/postfix-or-suffix

File: dialyzer.ignore-warnings.ee [hunk 6]
```diff
@@ -1,7 +1,8 @@
 # Errors
-riak_cs_block_server.erl:312: The pattern Success = {'ok', _} can never match the type {'error',_}
-riak_cs_block_server.erl:347: The pattern {'ok', RiakObject} can never match the type {'error',_}
-riak_cs_pbc.erl:100: The pattern {'ok', ClusterID} can never match the type {'error',_}
+riak_cs_block_server.erl:324: The pattern Success = {'ok', _} can never match the type {'error',_}
+riak_cs_block_server.erl:361: The pattern {'ok', RiakObject} can never match the type {'error',_}
+riak_cs_pbc.erl:58: The variable _ can never match since previous clauses completely covered the type 'pong'
```
Comment (kuenishi) [inline dialyzer.ignore-warnings.ee:4]: Note: https://github.com/basho/riak-erlang-client/issues/219

File: src/riak_cs_riak_client.erl (hunk 5 above)
Comment (kuenishi) [inline src/riak_cs_riak_client.erl]: Haha > "In this regard it is very much like prepone in Indian English"

Comment (kuenishi) [inline src/riak_cs_riak_client.erl]: I prefer suffix, to avoid confusion with a mailer daemon.

**Comment by kuenishi** [comment]
Code looks nice, a lot simpler than I thought. Are you gonna update this branch? Or I'll just +1 now and go to next round.

**Comment by kuenishi** [comment]
riak_test and manual tests have passed.

File: dialyzer.ignore-warnings.ee (hunk 6 above)
Comment (shino) [inline dialyzer.ignore-warnings.ee:4]: It is still open. Should be reviewed.

File: src/riak_cs_block_server.erl (hunk 1 above)
Comment (shino) [inline src/riak_cs_block_server.erl:318]: I won't make change about this.

**Comment by kuenishi** [comment]
Won't need fix for this right now, but Lifecycle mock fails like this:

```
DEBUG: format_uri(): http://aaa.s3.amazonaws.com/?lifecycle
DEBUG: Sending request method_string='GET', uri='http://aaa.s3.amazonaws.com/?lifecycle', headers={'Authorization': 'AWS admin-key:IpJege
EZQC8J39r0D9gnjVMwEFM=', 'x-amz-date': 'Thu, 16 Jul 2015 05:34:30 +0000'}, body=(0 bytes)
DEBUG: Response: {'status': 403, 'headers': {'date': 'Thu, 16 Jul 2015 05:34:31 GMT', 'content-length': '158', 'content-type': 'applicati
on/xml', 'server': 'Riak CS'}, 'reason': 'Forbidden', 'data': '<?xml version="1.0" encoding="UTF-8"?><Error><Code>AccessDenied</Code><Mes
sage>Access Denied</Message><Resource>/aaa</Resource><RequestId></RequestId></Error>'}
DEBUG: ConnMan.put(): closing proxy connection (keep-alive not yet supported)
ERROR: S3 error: Access Denied
   Expiration Rule: none
```

File: src/riak_cs_gc.erl (hunk 2 above)
Comment (shino) [inline src/riak_cs_gc.erl:384]: :pray:

File: src/riak_cs_riak_client.erl (hunk 5 above)
Comment (shino) [inline src/riak_cs_riak_client.erl]: As you wish :crown:

**Comment by shino** [comment]
Addressed your comments and pushed. Thanks for review, it helps much :smile:

**Comment by shino** [comment]
Ah, I don't look at location endpoint error. Will you file it as a new issue?

**Comment by shino** [comment]
Could not reproduce 403 for lifecycle endpoint in straightforward manner. My s3curl (I guess everyone owns his/her own s3curl, right?) does not understand `?lifecycle` as one of subresources and riak cs does not either. If `lifecycle` is added to s3curl for string to sign, them I had 403.

h/t to @ksauzz  for pointing out subresource mismatch for 403 error

**Comment by kuenishi** [comment]
My laptop eunit failed:

```
$ ./rebar eunit skip_deps=true suites=riak_cs_list_objects_fsm_v2_eqc
(snip)
........................................................x................................................................................
.........................................................................................................................................
.....................x........................
OK, passed 1000 tests
riak_cs_list_objects_fsm_v2_eqc:60: eqc_test_...*failed*
in function eqc:quickcheck/1 (../src/eqc.erl, line 1160)
in call from riak_cs_list_objects_fsm_v2_eqc:'-eqc_test_/0-fun-9-'/0 (test/riak_cs_list_objects_fsm_v2_eqc.erl, line 60)
**exit:{badarg,
    [{ets,lookup,[exometer_1,[riak_cs,in,riakc,fold_manifest_objs]],[]},
     {exometer,update_,2,[{file,"src/exometer.erl"},{line,213}]},
     {riak_cs_stats,safe_update,2,[{file,"src/riak_cs_stats.erl"},{line,220}]},
     {riak_cs_list_objects_fsm_v2,update_profiling_state_with_start,4,
         [{file,"src/riak_cs_list_objects_fsm_v2.erl"},{line,570}]},
     {riak_cs_list_objects_fsm_v2,make_2i_request,2,
         [{file,"src/riak_cs_list_objects_fsm_v2.erl"},{line,360}]},
     {riak_cs_list_objects_fsm_v2,prepare,2,
         [{file,"src/riak_cs_list_objects_fsm_v2.erl"},{line,138}]},
     {gen_fsm,handle_msg,7,[{file,"gen_fsm.erl"},{line,505}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,239}]}]}


riak_cs_list_objects_fsm_v2_eqc:64: eqc_test_...*failed*
in function eqc:quickcheck/1 (../src/eqc.erl, line 1160)
in call from riak_cs_list_objects_fsm_v2_eqc:'-eqc_test_/0-fun-12-'/0 (test/riak_cs_list_objects_fsm_v2_eqc.erl, line 64)
**exit:{badarg,
    [{ets,lookup,[exometer_2,[riak_cs,in,riakc,fold_manifest_objs]],[]},
     {exometer,update_,2,[{file,"src/exometer.erl"},{line,213}]},
     {riak_cs_stats,safe_update,2,[{file,"src/riak_cs_stats.erl"},{line,220}]},
     {riak_cs_list_objects_fsm_v2,update_profiling_state_with_start,4,
         [{file,"src/riak_cs_list_objects_fsm_v2.erl"},{line,570}]},
     {riak_cs_list_objects_fsm_v2,make_2i_request,2,
         [{file,"src/riak_cs_list_objects_fsm_v2.erl"},{line,360}]},
     {riak_cs_list_objects_fsm_v2,prepare,2,
         [{file,"src/riak_cs_list_objects_fsm_v2.erl"},{line,138}]},
     {gen_fsm,handle_msg,7,[{file,"gen_fsm.erl"},{line,505}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,239}]}]}


=======================================================
  Failed: 2.  Skipped: 0.  Passed: 2.
ERROR: One or more eunit tests failed.
ERROR: eunit failed while processing /home/kuenishi/cs-2.1/riak_cs: rebar_abort
```

**Comment by kuenishi** [comment]
@borshop merge

**Comment by Basho-JIRA** [comment]
This is a part of RCS-11 work; release notes will be written there. 

**_[posted via JIRA by Kota Uenishi]**_
